src/components/calcite-dropdown/calcite-dropdown.tsx

```diff
@@ -0,0 +1,216 @@
+    if (e.target.slot === "dropdown-trigger") {
+      if (
+        e.target.nodeName !== "BUTTON" &&
+        e.target.nodeName !== "CALCITE-BUTTON"
+      ) {
+        switch (e.key) {
+          case " ":
+          case "Enter":
+            this.openCalciteDropdown();
+            break;
+          case "Escape":
+            this.closeCalciteDropdown();
+            break;
+        }
+      } else if (e.key === "Escape" || (e.shiftKey && e.key === "Tab")) {
+        this.closeCalciteDropdown();
+      }
+    }
+  }
+
+  @Listen("calciteDropdownItemKeyEvent") calciteDropdownItemKeyEvent(
+    item: CustomEvent
+  ) {
+    let e = item.detail.item;
+    let isFirstItem = this.itemIndex(e.target) === 0;
+    let isLastItem = this.itemIndex(e.target) === this.items.length - 1;
+    switch (e.key) {
+      case "Tab":
+        if (isLastItem && !e.shiftKey) this.closeCalciteDropdown();
+        if (isFirstItem && e.shiftKey) this.closeCalciteDropdown();
+        break;
+      case "ArrowDown":
+        this.focusNextItem(e.target);
+        break;
+      case "ArrowUp":
+        this.focusPrevItem(e.target);
+        break;
+      case "Home":
+        this.focusFirstItem();
+        break;
+      case "End":
+        this.focusLastItem();
+        break;
+    }
+  }
+
+  @Listen("registerCalciteDropdownGroup") registerCalciteDropdownGroup(
+    e: CustomEvent
+  ) {
+    const items = {
+      items: e.detail.items as HTMLCalciteDropdownItemElement,
+      position: e.detail.position
+    };
+    this.items.push(items);
+  }
+
+  //--------------------------------------------------------------------------
+  //
+  //  Private State/Props
+  //
+  //--------------------------------------------------------------------------
+
+  /** created list of dropdown items */
+  @State() private items = [];
+
+  @State() private sorted = false;
+
+  /** unique id for dropdown */
+  /** @internal */
+  private dropdownId = `calcite-dropdown-${guid()}`;
+
+  //--------------------------------------------------------------------------
+  //
+  //  Private Methods
+  //
+  //--------------------------------------------------------------------------
+
+  private closeCalciteDropdown() {
+    this.active = false;
+  }
+
+  private focusFirstItem() {
+    const firstItem = this.items[0];
+    firstItem.focus();
+  }
+
+  private focusLastItem() {
+    const lastItem = this.items[this.items.length - 1];
+    lastItem.focus();
+  }
+
+  private focusNextItem(e) {
+    const index = this.itemIndex(e);
+    const nextItem = this.items[index + 1] || this.items[0];
+    nextItem.focus();
+  }
+
+  private focusPrevItem(e) {
+    const index = this.itemIndex(e);
+    const prevItem = this.items[index - 1] || this.items[this.items.length - 1];
+    prevItem.focus();
+  }
+
+  private itemIndex(e) {
+    return this.items.indexOf(e);
+  }
+
+  private openCalciteDropdown() {
+    this.active = !this.active;
+    this.focusFirstItem();
+  }
+
+  private sortItems() {
```

Comment:
This is a good opportunity to create a pure function! Instead of a method `sortItems(): void` which mutates `this`, you could create a separate `sortItems(items: any[]): any[]` and use the result of it to do the mutation of `this`.

The advantage of this approach isn't always obvious but I find that as a practice it makes OO code more understandable.

``` 
someLifeCycleMethod() {
  this.items = this.sortItems(this.items);
  this.sorted = true;
}

private sortItems = (items: any[]): any[] => 
  items.sort((a, b) => a.position - b.position)
    .concat.apply([], this.items.map(item => item.items))

```

Now that `sortItems` doesn't really on anything other than the arguments passed to it, it doesn't even really have to be a part of this class! You could just define it in some utilities folder. Alternatively, if you think it still makes sense in for it to be attached to this class, you can make it a `static` method that doesn't even have access to `this`.

Reply:
But there's nothing wrong with the code as-is. I'm just sharing a technique that I enjoy :)

Reply:
Rad, restructured to match this. Didn't extract into an external utility since this sort, and the sort on `calcite-dropdown-group`, are each slightly different, and still rely on a likely unique-to-this-use-case "position". But, updated :)